src/ring_theory/dedekind_domain/adic_valuation.lean

```diff
@@ -332,4 +332,96 @@ instance adic_completion.has_lift_t : has_lift_t K (v.adic_completion K) :=
 /-- The ring of integers of `adic_completion`. -/
 def adic_completion_integers : valuation_subring (v.adic_completion K) := valued.v.valuation_subring
 
+instance : inhabited (adic_completion_integers K v) := ⟨0⟩
+
+variables (R K)
+
+lemma adic_completion.is_integer (x : v.adic_completion K) :
+  x ∈ v.adic_completion_integers K ↔ (valued.v x : ℤₘ₀) ≤ 1 :=
+iff.rfl
+
+section algebra_instances
+
+@[priority 100] instance adic_valued.has_uniform_continuous_const_smul' :
+  @has_uniform_continuous_const_smul R K v.adic_valued.to_uniform_space _ :=
+@has_uniform_continuous_const_smul_of_continuous_const_smul R K _ _ _
+    v.adic_valued.to_uniform_space _ _
+
+instance adic_valued.has_uniform_continuous_const_smul :
+  @has_uniform_continuous_const_smul K K v.adic_valued.to_uniform_space _ :=
+@ring.has_uniform_continuous_const_smul K _ v.adic_valued.to_uniform_space _ _
+
+instance adic_completion.algebra' : algebra R (v.adic_completion K) :=
+@uniform_space.completion.algebra K _ v.adic_valued.to_uniform_space _ _ R _ _
+  (adic_valued.has_uniform_continuous_const_smul' R K v)
+
+@[simp] lemma coe_smul_adic_completion (r : R) (x : K) :
+  (↑(r • x) : v.adic_completion K) = r • (↑x : v.adic_completion K) :=
+@uniform_space.completion.coe_smul R K v.adic_valued.to_uniform_space _ _ r x
+
+instance : algebra K (v.adic_completion K) :=
+@uniform_space.completion.algebra' K _ v.adic_valued.to_uniform_space _ _
```

Comment:
Probably better to implement this as `adic_completion.algebra'`, assuming that works; it's less unfolding to unify the two algebra instances.

Reply:
I don't think that's right, this is not the general situation of any `R`-algebra `K` but rather the specific situation of the `R`-algebra of the field of fractions of `R`, and the completion is anchored at a place `v` of `R`.

Reply:
Huh, that seems not to work.

Reply:
This instance (which I'm surprised is missing) seemed to help a bit, but not enough
```
instance : is_fraction_ring K K :=
{ map_units := λ x, is_unit_of_mem_non_zero_divisors x.prop,
  surj := λ x, ⟨(x, 1), mul_one _⟩,
  eq_iff_exists := λ x y, ⟨λ h : x = y, ⟨1, h ▸ rfl⟩,
  λ ⟨c, h⟩, mul_left_cancel₀ (is_unit_of_mem_non_zero_divisors c.prop).ne_zero h⟩ }
```

Reply:
That instance is worth having independently, but I don't think it's a step in the right direction here: we are concerned with the completion of the field of fractions at a place _in the integers_.

Reply:
Hmm, in that case I retract my comment about not needing separate primed and unprimed smul lemmas.

Is there a generalization that gives us both algebra instances at once?

Reply:
> Hmm, in that case I retract my comment about not needing separate primed and unprimed smul lemmas.

I was aware of this but decided to let it go since I added lemmas characterising the `algebra_map`s :-)

> Is there a generalization that gives us both algebra instances at once?

Maybe if we introduced another set of coefficients `S` acting on `K` through `R`. I'm not sure. I think this is not worth fighting for.